### DISCUSSION TITLE
Ref usage in v7

### DISCUSSION THREAD

**Discussion by dan-lee** [react-hook-form/react-hook-form#5155] [Q&A] [answered]
In v6 I did the following in react-native and it worked great:

```tsx
<Controller
  control={control}
  name="name"
  rules={{ required: true }}
  render={({ onChange, onBlur, value, ref }) => (
    <Row
      onPress={() => ref.current?.focus()} // ← use the ref here
      title="Name"
      body={
        <TextInput
          ref={ref} // ← assign the ref here
          onBlur={() => {
            setValue('name', value.trim())
            onBlur()
          }}
          onChangeText={(text) => onChange(text)}
          value={value}
          {...otherProps}
        />
      }
    />
  )}
/>
```

I tried upgrading to v7 and changed the `render` callback to: 
```tsx
render={({ field: { onChange, onBlur, value, ref } }) => ( …
```

Now I am getting this TS error:
> TS2339: Property 'current' does not exist on type 'RefCallBack'.

I am wondering how a ref callback would work in this scenario

**Comment by bluebill1049** [comment]
can you share a codesandbox?

**Reply by dan-lee** [reply]
Sure:
- [CodeSandbox with working ref (v6)](https://codesandbox.io/s/wonderful-sinoussi-14r2y?file=/src/App.tsx)
- [CodeSandbox with failing ref (v7)](https://codesandbox.io/s/holy-flower-ko4dj?file=/src/App.tsx)

It's basically the same code, except for line 44.

In v6 you can click anywhere inside the green border and the input gets focused, with v7 it does not work anymore and you get the TS error mentioned above

**Reply by bluebill1049** [reply]
that's all you need to assign the `ref`

https://codesandbox.io/s/hopeful-noyce-jjggw?file=/src/App.tsx

**Reply by dan-lee** [reply]
Your example does not work for me.

In v6 the beauty of the code was, that I didn't need to assign the ref myself, I could just use it as `MutableRefObject`, now with the v7 way I would need to have an extra variable outside `Controller` which would defeat the purpose? With that, I can just have my own ref in my component

**Comment by leapful** [comment, marked as answer]
@dan-lee 

If your use case is related to "focus", you can use "setFocus" from "useForm" hook to set focus field manually just like you did with ref in v6.

```
onPress={() => setFocus('name')}
```

Working sample: https://codesandbox.io/s/cool-curie-lfffk

**Reply by dan-lee** [reply]
Very cool, didn't see that before. Thanks a lot
